test/string_util_gtest.cc

```diff
@@ -1,10 +1,11 @@
 //===---------------------------------------------------------------------===//
-// statistics_test - Unit tests for src/statistics.cc
+// string_util_test - Unit tests for src/string_util.cc
 //===---------------------------------------------------------------------===//
 
 #include <tuple>
 
 #include "../src/internal_macros.h"
+#include "../src/string_util.cc"
```

Comment:
you shouldn't need to do this, ever.  if you need to test a function, make it part of the public API.  but ideally, test the public API instead.

Reply:
Could you help me a little in this as I am not sure how to do this public API thing.

Reply:
only call functions in tests that are in the header (.h) file.  ie, you test against the public API (which is defined in the header) and not the private implementation (from the .cc file).

Reply:
These are the test cases -
```cpp
TEST(StringUtilTest, AppendHumanReadable) {
  std::string str;

  benchmark::AppendHumanReadable(0, &str);
  EXPECT_EQ("0", str);

  benchmark::AppendHumanReadable(999, &str);
  EXPECT_EQ("999", str);

  benchmark::AppendHumanReadable(1000, &str);
  EXPECT_EQ("1.00k", str);

  benchmark::AppendHumanReadable(1000000, &str);
  EXPECT_EQ("1.00M", str);

  benchmark::AppendHumanReadable(1000000000, &str);
  EXPECT_EQ("1.00G", str);
}

TEST(StringUtilTest, HumanReadableNumber) {
  EXPECT_EQ("1.00", benchmark::HumanReadableNumber(1.0));
  EXPECT_EQ("1.23Ki", benchmark::HumanReadableNumber(1234.0));
  EXPECT_EQ("1.00Mi", benchmark::HumanReadableNumber(1.0e6));
  EXPECT_EQ("1.00Gi", benchmark::HumanReadableNumber(1.0e9));

  EXPECT_EQ("1.00",
            benchmark::HumanReadableNumber(1.0, benchmark::Counter::kIs1000));
  EXPECT_EQ("1.23k", benchmark::HumanReadableNumber(
                         1234.0, benchmark::Counter::kIs1000));
  EXPECT_EQ("1000M",
            benchmark::HumanReadableNumber(1.0e6, benchmark::Counter::kIs1000));
  EXPECT_EQ("1000G",
            benchmark::HumanReadableNumber(1.0e9, benchmark::Counter::kIs1000));
}
```

and the output now - 
```bash
➜  test git:(inconsistent_suffixes_console_reporter_1009) ✗ ./string_util_gtest 
Running main() from gmock_main.cc
[==========] Running 6 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 6 tests from StringUtilTest
[ RUN      ] StringUtilTest.stoul
[       OK ] StringUtilTest.stoul (0 ms)
[ RUN      ] StringUtilTest.stoi
[       OK ] StringUtilTest.stoi (0 ms)
[ RUN      ] StringUtilTest.stod
[       OK ] StringUtilTest.stod (0 ms)
[ RUN      ] StringUtilTest.StrSplit
[       OK ] StringUtilTest.StrSplit (0 ms)
[ RUN      ] StringUtilTest.AppendHumanReadable
/home/devansh/benchmark/test/string_util_gtest.cc:170: Failure
Expected equality of these values:
  "999"
  str
    Which is: "0999"
/home/devansh/benchmark/test/string_util_gtest.cc:173: Failure
Expected equality of these values:
  "1.00k"
  str
    Which is: "09991000"
/home/devansh/benchmark/test/string_util_gtest.cc:176: Failure
Expected equality of these values:
  "1.00M"
  str
    Which is: "09991000976.562Ki"
/home/devansh/benchmark/test/string_util_gtest.cc:179: Failure
Expected equality of these values:
  "1.00G"
  str
    Which is: "09991000976.562Ki953.674Mi"
[  FAILED  ] StringUtilTest.AppendHumanReadable (0 ms)
[ RUN      ] StringUtilTest.HumanReadableNumber
/home/devansh/benchmark/test/string_util_gtest.cc:183: Failure
Expected equality of these values:
  "1.00"
  benchmark::HumanReadableNumber(1.0)
    Which is: "1"
/home/devansh/benchmark/test/string_util_gtest.cc:184: Failure
Expected equality of these values:
  "1.23Ki"
  benchmark::HumanReadableNumber(1234.0)
    Which is: "1.20508Ki"
/home/devansh/benchmark/test/string_util_gtest.cc:185: Failure
Expected equality of these values:
  "1.00Mi"
  benchmark::HumanReadableNumber(1.0e6)
    Which is: "976.562Ki"
/home/devansh/benchmark/test/string_util_gtest.cc:186: Failure
Expected equality of these values:
  "1.00Gi"
  benchmark::HumanReadableNumber(1.0e9)
    Which is: "953.674Mi"
/home/devansh/benchmark/test/string_util_gtest.cc:188: Failure
Expected equality of these values:
  "1.00"
  benchmark::HumanReadableNumber(1.0, benchmark::Counter::kIs1000)
    Which is: "1"
/home/devansh/benchmark/test/string_util_gtest.cc:190: Failure
Expected equality of these values:
  "1.23k"
  benchmark::HumanReadableNumber( 1234.0, benchmark::Counter::kIs1000)
    Which is: "1.234k"
/home/devansh/benchmark/test/string_util_gtest.cc:192: Failure
Expected equality of these values:
  "1000M"
  benchmark::HumanReadableNumber(1.0e6, benchmark::Counter::kIs1000)
    Which is: "1000k"
/home/devansh/benchmark/test/string_util_gtest.cc:194: Failure
Expected equality of these values:
  "1000G"
  benchmark::HumanReadableNumber(1.0e9, benchmark::Counter::kIs1000)
    Which is: "1000M"
[  FAILED  ] StringUtilTest.HumanReadableNumber (0 ms)
[----------] 6 tests from StringUtilTest (1 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 1 test suite ran. (1 ms total)
[  PASSED  ] 4 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] StringUtilTest.AppendHumanReadable
[  FAILED  ] StringUtilTest.HumanReadableNumber

 2 FAILED TESTS
```